### PR TITLE
fix order() to handle complex subqueries

### PR DESCRIPTION
**Pull request by kbrock**
**TL;DR:** remove sql string munging for `ORDER BY` and leverage Arel instead.

### Summary

If a query has a `distinct()` and an `order()`, active record generates a pre query to just fetch ids. `columns_for_distinct` converts the order clause into something usable in the sql clause.

```ruby
Author.distinct.left_joins(:posts).where(:posts => {:id => [5,6,7,8,9,10]}).order(:name)
```

```sql
SELECT DISTINCT "authors"."id" AS alias_0, "authors"."name"
LEFT OUTER JOIN "posts" ON "posts"."author_id" = "authors"."id"
FROM "authors"
WHERE "posts"."id" in (...)
ORDER BY "posts"."name";
```

The issue enters when the `order()` clause is a subquery joining to a `has_many` model. The subquery needs to return 1 value, so it will have a `LIMIT 1` and an `ORDER BY` clause.

The sql munging in `columns_for_distinct` inadvertently modifies the subquery's `ORDER BY` clause. So the `SELECT` subquery will be different from the `ORDER BY` subquery. Postgres and MySQL do not like this:

```
ActiveRecord::StatementInvalid (PG::InvalidColumnReference: ERROR:  for SELECT DISTINCT, ORDER BY expressions must appear in select list)
LINE 1: ..."disks"."hardware_id" = "hardwares"."id" ORDER BY (SELECT  "...
```

### Before this PR

- Rails converts the `ORDER BY` clause columns to a sql string and uses regular expressions to removes  all `"ASC", "DESC"` from the sql.
- The sql in the `SELECT` no longer resembles the sql in the `ORDER BY` and the db complains.

### After

- Rails drops the `Arel::Nodes::Ordering` node and uses the actual expression (i.e.: `@expr`). Since the `Order` node is removed, the `ASC` or `DESC` are successfully removed. No String munging required.
- To support `Arel::Attributes::Attribute` nodes, the general case of `to_sql` is added.

I've observed that most of the `ORDER BY` clauses are an `Arel::Nodes::Ordering` node pointing to an `Arel::Attributes::Attribute` node. Less frequently, the `ORDER BY` clauses is an `Ordering` node pointing to a `SqlLiteral`. This supports both.

### Alternatives

- ~~I can implement `Arel::Attributes::Attribute#to_sql`, but comments suggests that `to_sql` is bad and should go away.~~ **DONE**
- ~~I can move `columns_for_distinct` into the abstract `SchemaStatements`, if you prefer.~~ **DONE**

### Thanks

Thank for all the help.

As @matthewd and @tenderlove may have guessed, this is related to [virtual attributes](https://github.com/ManageIQ/activerecord-virtual_attributes).

### CODE REVIEW

**Review by d-m-u — APPROVED** [review]

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb [hunk 1]
```diff
@@ -557,10 +557,25 @@ def type_to_sql(type, limit: nil, precision: nil, scale: nil, array: nil, **) #
...
           order_columns = orders.reject(&:blank?).map { |s|
               # Convert Arel node to string
-              s = s.to_sql unless s.is_a?(String)
+              unless s.is_a?(String)
+                # if this is an order node, then just use the core contents without modification
+                if s.kind_of?(Arel::Nodes::Ordering)
+                  s = s.expr
+                  keep_order = true
+                end
+                # essentially s = s.to_sql but works for attributes too
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb]: This would be better handled using polymorphism by calling `to_sql` and it working for any class. `to_sql` should not be used by apps but rails can use it.

**Review by rafaelfranca — COMMENTED** [review]
Is this problem specific for postgresql or can happen on mysql/sqlite too?

**Comment by kbrock** [comment]
thanks @rafaelfranca - let me know if you want me to squash the to_sql into the original commit (think your rules are to NOT do that, but I thought I'd ask)

Also, you want the specs moved as well?

**Comment by kbrock** [comment]
rebased to fix merge conflict.

rearranged commits to better partition what is done:

1. Generalize `columns_for_distinct` (and tests) per @rafaelfranca request
2. Define `to_sql` upfront.
3. Change `columns_for_distinct` implementation to rely less upon regex munging.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb [hunk 2]
```diff
@@ -1148,7 +1148,26 @@ def type_to_sql(type, limit: nil, precision: nil, scale: nil, **) # :nodoc:
       #   columns_for_distinct("posts.id", ["posts.created_at desc"])
       #
       def columns_for_distinct(columns, orders) # :nodoc:
-        columns
+        order_columns = orders.compact_blank.map { |s|
```
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: I'm curious why returning order_columns for all adapters?
AFAIK this is not necessary for SQLite3.

Comment (kbrock) [reply activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: AFAIK, this is ANSI SQL standard.

I undid for SQLite3 to be simpler.

Comment (kbrock) [reply activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: AFAIK, this is ANSI SQL standard.

I undid for SQLite3 to be simpler.

**Review by kbrock — COMMENTED** [review]
Now that SQLite3 is returned to the original, is this closer to what you want?

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb (hunk 2 above)
Comment (kbrock) [reply activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: @kamipo I have fixed to return simple query for SQLite3

Since this is typical behavior for ANSI SQL standard, I thought it better to make default behavior rather than override for mysql, postgres (and probably others)

Does this change fix your issues?

Comment (kbrock) [reply activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:1152]: @kamipo are you ok with this pr now?

anything else that would help you understand / merge?

**Comment by kbrock** [comment]
@rafaelfranca I moved the specs to generic connection handler.
does this make it too complicated for you - would you prefer that I make this PR smaller and just special case for mysql / postgres?
